codebuild/bin/s2n_open_fds_test.py

```diff
@@ -0,0 +1,62 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+# This script parses the LastDynamicAnalysis file generated by Valgrind running through CTest memcheck.
+# It identifies any leaking file descriptors and triggers an error when detected.
+# This enhances the capabilities of existing Valgrind checks.
+# Output snippet for open file descriptors:
+#  ==6652== FILE DESCRIPTORS: 6 open (3 std) at exit.
+#  ==6652== Open AF_INET socket 6: 127.0.0.1:36915 <-> unbound
+#  ==6652==    at 0x498B2EB: socket (syscall-template.S:120)
+#  ==6652==    by 0x16CD16: s2n_new_inet_socket_pair (s2n_self_talk_ktls_test.c:69)
+#  ==6652==    by 0x15DBB2: main (s2n_self_talk_ktls_test.c:168)
+#  ==6652==
+import os
+import sys
+
+EXIT_SUCCESS = 0
+# Exit with error code 1 if leaking fds are detected.
+ERROR_EXIT_CODE = 1
+# This test is designed to be informational only, so we only print fifteen lines of error messages when a leak is detected.
+NUM_OF_LINES_TO_PRINT = 15
+
+
+def detect_leak(file):
+    fd_leak_detected = False
+    lines = file.readlines()
+    for i in range(len(lines)):
+        if "FILE DESCRIPTORS:" in lines[i]:
+            # Example line: `==6096== FILE DESCRIPTORS: 4 open (3 std) at exit.`
+            line_elements = lines[i].split()
+            open_fd_count = line_elements[line_elements.index("DESCRIPTORS:") + 1]
+            std_fd_count = line_elements[line_elements.index("std)") - 1][1:]
+            # CTest memcheck writes to a LastDynamicAnslysis log file.
+            # We allow that fd to remain opened.
+            if int(open_fd_count) - int(std_fd_count) > 1:
+                for j in range(NUM_OF_LINES_TO_PRINT):
+                    print(lines[i + j], end="")
+                print()
+                fd_leak_detected = True
+    return fd_leak_detected
+
+
+def main():
+    # Print banner of the test
+    print("############################################################################")
+    print("################# Test for Leaking File Descriptors ########################")
+    print("############################################################################")
+
+    path = sys.argv[1]
+    for f in os.listdir(path):
+        if "LastDynamicAnalysis" in f:
+            with open(os.path.join(path, f), 'r') as file:
+                if detect_leak(file):
+                    sys.exit(ERROR_EXIT_CODE)
+                else:
+                    return EXIT_SUCCESS
```

Comment:
What do you think about abstracting this out into a separate "find_file" function?

That way
- error handling is consolidated
- the program flow is more linear instead of "horizontal"
    - 5 level of indentation is quite a bit.

```suggestion
    log_file_path = find_log_file(sys.argv[1])
    with open(log_file_path, 'r') as file:
        if detect_leak(file):
            sys.exit(ERROR_EXIT_CODE)
```

Reply:
I agree with your suggestions and have implemented them. I would also say that we don't need `log_file_path` intermediate variable, we can just call that function with `open`.